packages/pnpm-sync-lib/src/pnpmSyncPrepare.ts

```diff
@@ -49,15 +63,106 @@ export interface IPnpmSyncPrepareOptions {
-   *
-   * @remarks
-   * `LogMessageKind.ERROR` events do NOT cause the promise to reject,
-   * so they must be handled appropriately.
+   * The root folder path of the project.
    */
-  logMessageCallback: (options: ILogMessageCallbackOptions) => void;
+  projectFolder: string;
+
+  /**
+   * A set of target folders that need to be synchronized.
+   */
+  targetFolderSet: Set<string>;
+}
+
+/**
+ * @beta
+ */
+export async function writePnpmSyncFileAsync(options: ICreatePnpmSyncOptions): Promise<void> {
+  const { lockfileId, logMessageCallback, projectFolder, targetFolderSet } = options;
+
+  const pnpmSyncJsonFolder = `${projectFolder}/node_modules`;
+  const pnpmSyncJsonPath = `${pnpmSyncJsonFolder}/.pnpm-sync.json`;
+  const expectedPnpmSyncJsonVersion: string = pnpmSyncGetJsonVersion();
+
+  let pnpmSyncJsonFile: IPnpmSyncJson = {
+    version: expectedPnpmSyncJsonVersion,
+    postbuildInjectedCopy: {
+      sourceFolder: '..', // path from pnpmSyncJsonFolder to projectFolder
+      targetFolders: []
+    }
+  };
+
+  // if .pnpm-sync.json already exists, read it first
+  if (fs.existsSync(pnpmSyncJsonPath)) {
+    let existingPnpmSyncJsonFile: IPnpmSyncJson | undefined;
+    try {
+      existingPnpmSyncJsonFile = JSON.parse(fs.readFileSync(pnpmSyncJsonPath).toString());
+    } catch (e) {
+      // no-catch
+      // Regenerate .pnpm-sync.json when failed to load the current one
+    }
+
+    if (existingPnpmSyncJsonFile) {
+      const actualPnpmSyncJsonVersion: string = existingPnpmSyncJsonFile.version;
+      if (actualPnpmSyncJsonVersion === expectedPnpmSyncJsonVersion) {
+        // If a lockfileId is provided
+        // then all entries with this lockfileId should be deleted
+        // they will be regenerated later
+        if (lockfileId) {
+          const filteredTargetFolders = existingPnpmSyncJsonFile.postbuildInjectedCopy.targetFolders.filter(
+            (targetFolder) => targetFolder?.lockfileId !== lockfileId
+          );
+          existingPnpmSyncJsonFile.postbuildInjectedCopy.targetFolders = filteredTargetFolders;
+        }
+        pnpmSyncJsonFile = existingPnpmSyncJsonFile;
+      } else {
+        logMessageCallback({
+          message: `The .pnpm-sync.json file in ${pnpmSyncJsonFolder} has an incompatible version; pnpm-sync will regenerate it.`,
+          messageKind: LogMessageKind.VERBOSE,
+          details: {
+            messageIdentifier: LogMessageIdentifier.PREPARE_REPLACING_FILE,
+            pnpmSyncJsonPath,
+            projectFolder,
+            actualVersion: actualPnpmSyncJsonVersion,
+            expectedVersion: expectedPnpmSyncJsonVersion
+          }
+        });
+      }
+    }
+  }
+
+  const existingTargetFolderSet: Set<string> = new Set();
+
+  for (const targetFolder of pnpmSyncJsonFile.postbuildInjectedCopy.targetFolders) {
+    existingTargetFolderSet.add(targetFolder.folderPath);
+  }
+
+  for (const targetFolder of targetFolderSet) {
```

Comment:
This function doesn't seem to use any functionality of `Set<string>`.  It simply treats it like a list.

The caller of the function will have an easier time if we can accept an array or maybe iterator or something that doesn't force the caller to make a `Set`.